scripts/ci-report/main.go

```diff
@@ -0,0 +1,188 @@
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+	"time"
+
+	"golang.org/x/exp/slices"
+)
+
+func main() {
+	if len(os.Args) != 2 {
+		_, _ = fmt.Println("usage: ci-report <gotests.json>")
+		os.Exit(1)
+	}
+	name := os.Args[1]
+
+	f, err := os.Open(name)
+	if err != nil {
+		_, _ = fmt.Printf("error opening gotestsum json file: %v", err)
+		os.Exit(1)
+	}
+	defer f.Close()
+
+	dec := json.NewDecoder(f)
+	var report GotestsumReport
+	for {
+		var e GotestsumReportEntry
+		err = dec.Decode(&e)
+		if errors.Is(err, io.EOF) {
+			break
+		}
+		if err != nil {
+			_, _ = fmt.Printf("error decoding json: %v", err)
+			os.Exit(1)
+		}
+		e.Package = strings.TrimPrefix(e.Package, "github.com/coder/coder/")
+		report = append(report, e)
+	}
+
+	packagesSortedByName := []string{}
+	packageTimes := map[string]float64{}
+	packageFail := map[string]int{}
+	packageSkip := map[string]bool{}
+	testTimes := map[string]float64{}
+	testSkip := map[string]bool{}
+	testOutput := map[string]string{}
+	testSortedByName := []string{}
+	for i, e := range report {
+		switch e.Action {
+		// A package/test may fail or pass.
+		case Fail:
+			if e.Test == "" {
+				packageTimes[e.Package] = *e.Elapsed
+			} else {
+				packageFail[e.Package]++
+				name := e.Package + "." + e.Test
+				testTimes[name] = *e.Elapsed
+			}
+		case Pass:
+			if e.Test == "" {
+				packageTimes[e.Package] = *e.Elapsed
+			} else {
+				name := e.Package + "." + e.Test
+				delete(testOutput, name)
+				testTimes[name] = *e.Elapsed
+			}
+
+		// Gather all output (deleted when irrelevant).
+		case Output:
+			if e.Test != "" {
+				name := e.Package + "." + e.Test
+				testOutput[name] += e.Output
+			}
+
+		// Packages start, tests run and either may be skipped.
+		case Start:
+			packagesSortedByName = append(packagesSortedByName, e.Package)
+		case Run:
+			name := e.Package + "." + e.Test
+			testSortedByName = append(testSortedByName, name)
+		case Skip:
+			if e.Test == "" {
+				packageSkip[e.Package] = true
+			} else {
+				name := e.Package + "." + e.Test
+				testSkip[name] = true
+				delete(testOutput, name)
+			}
+
+		// Ignore.
+		case Cont:
+		case Pause:
+
+		default:
+			_, _ = fmt.Printf("unknown action: %v in entry %d (%v)", e.Action, i, e)
+			os.Exit(1)
+		}
+	}
+
+	sortAZ := func(a, b string) bool { return a < b }
+	slices.SortFunc(packagesSortedByName, sortAZ)
+	slices.SortFunc(testSortedByName, sortAZ)
```

Comment:
It's hard to review this PR without any reference to `gotests.json`. Do you think that you could record a short snippet, add it to the repo, and run a basic test?

Reply:
Added a sample and golden file, thoughts?

Reply:
Can you strip it a bit :) ?

```
21,327 additions, 0 deletions not shown because the diff is too large. Please use a local Git client to view these changes.
```

Reply:
I'd have to write a parser to strip it, can't just willy-nilly delete lines since it's a state-machine. 😅

Reply:
I guess it could be run on a subset of tests 🤔, but I'd like there to be failures to capture real-world.

Reply:
Maybe try to do this as it will add a 34MB file to the repository.

Reply:
Haha, good catch. Well, 852K is as low as I'll go. 😆